From 39284068f11ab95babbc0c4044c68092abce6c31 Mon Sep 17 00:00:00 2001
From: wuzhouhui <wuzhouhui250@gmail.com>
Date: Sun, 23 Oct 2016 20:13:36 +0800
Subject: [PATCH 04/95] Open .svnignore if it existed

We use apr_file_open() to open .svnignore, so change the type of svn_ignore_fd
from int svn_ignore_fd to apr_file_t *svn_ignore_file.
---
 subversion/include/private/svn_sqlite.h |  3 +++
 subversion/libsvn_subr/sqlite.c         | 13 ++++++++++++-
 subversion/libsvn_wc/wc_db_util.c       |  9 ++++++++-
 subversion/libsvn_wc/wc_db_wcroot.c     |  2 --
 4 files changed, 23 insertions(+), 4 deletions(-)

diff --git a/subversion/include/private/svn_sqlite.h b/subversion/include/private/svn_sqlite.h
index cd01348..877c5fc 100644
--- a/subversion/include/private/svn_sqlite.h
+++ b/subversion/include/private/svn_sqlite.h
@@ -129,6 +129,9 @@ svn_sqlite__open(svn_sqlite__db_t **db, const char *path,
                  apr_int32_t timeout,
                  apr_pool_t *result_pool, apr_pool_t *scratch_pool);
 
+svn_error_t *
+svn_sqlite__open_file(svn_sqlite__db_t *, const char *, apr_pool_t *);
+
 /* Explicitly close the connection in DB. */
 svn_error_t *
 svn_sqlite__close(svn_sqlite__db_t *db);
diff --git a/subversion/libsvn_subr/sqlite.c b/subversion/libsvn_subr/sqlite.c
index 9193209..fed2a6a 100644
--- a/subversion/libsvn_subr/sqlite.c
+++ b/subversion/libsvn_subr/sqlite.c
@@ -146,7 +146,7 @@ struct svn_sqlite__db_t
    * File descriptor of SVN_IGNORE_FILE. The role of SVN_IGNORE_FILE is
    * analogous to sqlite db, so I put it in this struct.
    */
-  int   svn_ignore_fd;
+  apr_file_t   *svn_ignore_file;
 
 #ifdef SVN_UNICODE_NORMALIZATION_FIXES
   /* Buffers for SQLite extensoins. */
@@ -1076,6 +1076,17 @@ like_ucs_nfd(sqlite3_context *context,
 #endif /* SVN_UNICODE_NORMALIZATION_FIXES */
 
 svn_error_t *
+svn_sqlite__open_file(svn_sqlite__db_t *db, const char *path, apr_pool_t *pool)
+{
+  apr_status_t ret;
+  ret = apr_file_open(&db->svn_ignore_file, path, APR_READ, 0, pool);
+  if (ret == APR_SUCCESS)
+    return SVN_NO_ERROR;
+  else
+    return svn_error_wrap_apr(ret, NULL);
+}
+
+svn_error_t *
 svn_sqlite__open(svn_sqlite__db_t **db, const char *path,
                  svn_sqlite__mode_t mode, const char * const statements[],
                  int unused1, const char * const *unused2,
diff --git a/subversion/libsvn_wc/wc_db_util.c b/subversion/libsvn_wc/wc_db_util.c
index 8402c42..d58c5f7 100644
--- a/subversion/libsvn_wc/wc_db_util.c
+++ b/subversion/libsvn_wc/wc_db_util.c
@@ -49,6 +49,8 @@
 
 WC_QUERIES_SQL_DECLARE_STATEMENTS(statements);
 
+#define SVN_IGNORE_FILE ".svnignore"
+
 
 
 /* */
@@ -122,10 +124,12 @@ svn_wc__db_util_open_db(svn_sqlite__db_t **sdb,
 {
   const char *sdb_abspath = svn_wc__adm_child(dir_abspath, sdb_fname,
                                               scratch_pool);
+  const char *ign_abspath = svn_dirent_join_many(scratch_pool, dir_abspath,
+		  SVN_IGNORE_FILE, SVN_VA_NULL);
+  svn_node_kind_t kind;
 
   if (smode != svn_sqlite__mode_rwcreate)
     {
-      svn_node_kind_t kind;
 
       /* A file stat is much cheaper than a failed database open handled
          by SQLite. */
@@ -141,6 +145,9 @@ svn_wc__db_util_open_db(svn_sqlite__db_t **sdb,
   SVN_ERR(svn_sqlite__open(sdb, sdb_abspath, smode,
                            my_statements ? my_statements : statements,
                            0, NULL, timeout, result_pool, scratch_pool));
+  SVN_ERR(svn_io_check_path(ign_abspath, &kind, scratch_pool));
+  if (kind == svn_node_file)
+    SVN_ERR(svn_sqlite__open_file(*sdb, ign_abspath, result_pool));
 
   if (exclusive)
     SVN_ERR(svn_sqlite__exec_statements(*sdb, STMT_PRAGMA_LOCKING_MODE));
diff --git a/subversion/libsvn_wc/wc_db_wcroot.c b/subversion/libsvn_wc/wc_db_wcroot.c
index 8637732..1cfca3d 100644
--- a/subversion/libsvn_wc/wc_db_wcroot.c
+++ b/subversion/libsvn_wc/wc_db_wcroot.c
@@ -38,8 +38,6 @@
 
 #include "svn_private_config.h"
 
-#define SVN_IGNORE_FILE ".svnignore"
-
 /* ### Same values as wc_db.c */
 #define SDB_FILE  "wc.db"
 #define UNKNOWN_WC_ID ((apr_int64_t) -1)
-- 
1.8.3.1

